### ISSUE TITLE
Numix-Circle PPA not updating

### ISSUE THREAD

**Issue by dirtydancing** [numixproject/circle-core#2430]
Question: Has the PPA updating of Numix-Circle stopped again? The current PPA status for Numix-Circle dates from 16 July, thus from 10 days ago, although this should update on a daily basis in case of changes in git. And there have been changes to Numix-Circle on GitHub over the last days.

![numix-circle_ppa](https://cloud.githubusercontent.com/assets/7164227/8893674/9ca3e534-339a-11e5-9e71-a3e515825cd6.png)

**Comment by Foggalong** [comment]
Yeah, the problem with the PPA was never fixed. It's all caused by a bug in Launchpad: if someone adds a submodule to the git repo and then deletes it in a subsequent commit then updates break. The only solution is to either find which commit the submodule was added in and delete it, or to completely wipe the repo (loosing all history, contributors list, etc) and start again.

I remember that the person who added the submodule was a community contributor (i.e. not on the Numix owners GitHub team) and so the commit was submitted via pull request. It was then reverted in the subsequent pull. I've tried for several hours to find this commit to delete it but all to no avail :(

**Comment by dirtydancing** [comment]
Would it make sense to put out a post on Google+ to possibly get hold of the community contributor or to get help in locating that pull request?

**Comment by Foggalong** [comment]
Because I know the pull request was actually merged we know that whoever it was is actually on the contributor list for the project. Rather than several 1000s of Google+ peeps we know it's (almost definitely) a merge from one of the following:

* [ ] @akashdevaraju 
* [ ] @alexskc
* [ ] @alfredvaa
* [ ] @Algram
* [ ] @bernimoses
* [ ] @bil-elmoussaoui 
* [ ] @ChuckDaniels87
* [ ] @csmart
* [ ]  @dirtydancing
* [ ] @Eleuin
* [ ] @ElmoTheGrower
* [ ] @elthommo
* [ ] @flexiondotorg
* [ ] @Ghokun
* [ ] @gregorydk
* [ ] @harp37
* [ ] @hlechner
* [ ] @jhrdka 
* [ ] @l0okme
* [ ] @martimorta
* [ ] @m-e-h
* [ ] @mendhak
* [ ] @mikegioia
* [ ] @moukoublen
* [ ] @nathaneltitane
* [ ] @nejsan
* [ ] @palob 
* [ ] @rickinbsb
* [ ] @tech4david
* [ ] @tvgdb
* [ ] @xiehuc

Just a case of looking through each persons pull requests and seeing if they did it or getting a comment from them saying it wasn't them. As a disclaimer, the person who did this isn't in any trouble at all. We just need to find the commit so we can get the PPA working again :)

**Comment by palob** [comment]
[I confess](https://github.com/numixproject/numix-icon-theme-circle/commit/c221a7007acb4142cf8946627dbd9c51d89b4dcf)
Don't know how I managed to do that. AFAIK I accidently pushed an empty folder.

Has been reverted in https://github.com/numixproject/numix-icon-theme-circle/pull/1289/files.

**Comment by Foggalong** [comment]
Thanks <3 now I just need to work out exactly what needs doing to resolve this problem

**Comment by Foggalong** [comment]
Trying to put together a list of commits that need to be deleted from the git history:

* c221a7007acb4142cf8946627dbd9c51d89b4dcf
* 814156e5cf043e3e9484c0d3644530f666807fad
* d094db7d045ca7507b468a0ffa3c9d1330ed3dae
* 9c610a87e0ae850dda753c9238b85e4f6641a708
* 5d7e48aa4201d8667e4c80a4603ce93ba9225ec0

**Comment by palob** [comment]
Looks like I did this more often then once :grimacing:

**Comment by Foggalong** [comment]
I think it's just once on creation and then referenced again when being merged. We've ran into the problem that we can't just delete this 5 commits because they also affect other icons which have also been modified since. If anyone has any ideas it would be much appreciated :(

**Comment by palob** [comment]
A hacky solution but is it possible to remove all icons changed in the toxic commits from master, delete the commits and then readd the files?

**Comment by Foggalong** [comment]
Problem is that it flows down in a tree effecting 100s of icons :(

**Comment by palob** [comment]
So nuking all icons and adding them again later?

**Comment by Foggalong** [comment]
Problem then is that fs up our issue tracking, other peoples forks, and means we can't refer back to github history :(

**Comment by palob** [comment]
So you think it's not possible to get rid of the commits retroactively anyway?

**Comment by Foggalong** [comment]
That's technically what git revert for, but the Launchpad bug dictates that it needs to be wiped from the history not just reveresed. It also seems to be any commit which refers to the submodue, not just the ones which created/removed it.

**Comment by palob** [comment]
Is there any reference for the launchpad bug?

**Comment by mendhak** [comment]
That's quite the messy bug - on one hand you shouldn't have to rewrite history to do this, on the other hand LP shouldn't be stopping you from pushing.

I don't know how difficult it is, but what about pushing straight to Launchpad using something like Jenkins or Travis? It was the [simpler solution](https://github.com/semiosis/glusterfs-debian/issues/5) in this example.

**Comment by palob** [comment]
Is there a way to let GitHub do the cloning via [here](https://github.com/numixproject/numix-icon-theme-circle/archive/master.zip) and build the ``.deb`` from the zip?

**Comment by rickinbsb** [comment]
Is it possible to fix exactly what caused the problem as a new commit
instead of reverting?

Em dom, 26 de jul de 2015 14:47, palob <notifications@github.com> escreveu:

**Comment by palob** [comment]
Nope. As long as the original commit is in the git history the problem will occur.

**Comment by vysinsky** [comment]
Hi, is this solved? I get normal updates from ppa :)

**Comment by palob** [comment]
But only for Numix base I suppose?
In [ppa:numix/ppa](https://launchpad.net/~numix/+archive/ubuntu/ppa) Numix Circle still dates from 2015-07-16.

**Comment by vysinsky** [comment]
oh you are right :/

**Comment by wa4557** [comment]
Hmm until #2546 comes into play an idea: would a
`git reset --hard 5e296198b44d0763dcec659b4648fd123356e938` (in a new branch of course)
and afterwards `git cherry-pick` the "correct" commits
and finally something like 
`git cherry-pick 81a4529^..51f34d0` (to get all the commits afterwards) work?

Or rebasing. With that one can rewrite the history (I tried it but it seems that I'm not good enough in "git" to fully succeed)

**Comment by palob** [comment]
Cherrypicking from a fork after the reset?

**Comment by wa4557** [comment]
or from master in a branch

**Comment by Foggalong** [comment]
Does that work with commits that depend on other commits?

**Comment by wa4557** [comment]
Yes, If you add the exact not toxic files from the "toxic" parent commit in a commit  in between cherry-picking, it works..

**Comment by palob** [comment]
So ``git reset --hard`` 5e296198b44d0763dcec659b4648fd123356e938, commiting the non-toxic  icons from my toxic PR and then cherrypicking everything **after** 5d7e48aa4201d8667e4c80a4603ce93ba9225ec0?

Did you test it and it worked? No submodule in the revision history?

If it works, all we'll lose/mess up is GitHub commit references.

**Comment by wa4557** [comment]
I tried it and it worked kind of but I didn't fully understand what needs to be picked and what not, which resulted in a merge conflict, but I'm pretty confident that it should work when done right.

**Comment by palob** [comment]
Maybe a test git repo simulating our situation is the way to get it straight?

**Comment by wa4557** [comment]
I guess we can do that in one of our forks?

**Comment by palob** [comment]
Or a local git repo with just a text file manipulated in the commits.

**Comment by wa4557** [comment]
I guess you know more about that than I do :) But I agree a new repo would be the best because we could talk about the issue over there

**Comment by palob** [comment]
You can still host the repo on Github of course to get the same workflow. I've thought of the one-file repo instead of a fork of Numix-Circle so as to avoid the unwieldy amount of commits and files.

**Comment by vysinsky** [comment]
Hi, will this be solved somehow? How can I install icons from this repo instead of using ppa?

**Comment by Foggalong** [comment]
Yeah, we hope so :) in the mean time you can indeed use the icons from this repo

**Comment by wa4557** [comment]
I tried my approach several times but In the end I always got a merge conflict, that could not be resolved. I still think however that it should be possible @palob Did you try as well, any better results?

**Comment by wa4557** [comment]
I figured it out, it's actually pretty simple :)
Running this on master rewrites the history and removes every trace of the submodule without altering the commit  history.
`git filter-branch --force --index-filter 'git rm --cached --ignore-unmatch numix-icon-theme-circle' --prune-empty --tag-name-filter cat -- --all`

and afterwards a 
`git push --force origin master`

proof: https://github.com/wa4557/numix-icon-theme-circle/commits/ppa?page=48

**Comment by Foggalong** [comment]
....how?!

**Comment by wa4557** [comment]
filter-branch hoes through all the commits and rewrites them in such a way, that all traces of the unwanted file are deleted (and nothing else). I guess this was introduced to remove traces of sensible data, but is also handy in our case.

**Comment by Foggalong** [comment]
I might be missing something then, but which part of that comman refers to the submodue?

**Comment by wa4557** [comment]
`git rm --cached --ignore-unmatch numix-icon-theme-circle`

**Comment by Foggalong** [comment]
wasn't the submodule file something like `.gitsubmodule`? Can't remember off hand. Again though, just asking out of curiosity I'm no git expert

**Comment by wa4557** [comment]
Me neither, was just playing around. I guess the submodule was an empty `numix-icon-theme-circle` folder inside the `numix-icon-theme-circle` folder. But if the submodule was something like `.gitsubmodule` we can remove that as well

**Comment by Foggalong** [comment]
I can't seem to find that now soi I'm guessing you're right on this :) I'll run the command and we'll wait and see if the builds start working again!

**Comment by wa4557** [comment]
You might want to try this out in a branch, and replace the master afterwards, just in case...

**Comment by Foggalong** [comment]
nah, already done :P this was the output

```bash
Rewrite c221a7007acb4142cf8946627dbd9c51d89b4dcf (1013/2701)rm 'numix-icon-theme-circle'
Rewrite 814156e5cf043e3e9484c0d3644530f666807fad (1014/2701)rm 'numix-icon-theme-circle'
Rewrite d094db7d045ca7507b468a0ffa3c9d1330ed3dae (1015/2701)rm 'numix-icon-theme-circle'
Rewrite 7c47343f65f843f96be479c6fd2b0fa9a78b92fe (1071/2701)rm 'numix-icon-theme-circle'
Rewrite f56c15d207f3e76f893aa332a540632b6a196489 (2701/2701)
Ref 'refs/heads/master' was rewritten
Ref 'refs/remotes/origin/master' was rewritten
Ref 'refs/remotes/origin/hidpi' was rewritten
WARNING: Ref 'refs/remotes/origin/master' is unchanged
WARNING: Ref 'refs/tags/v1' is unchanged
v1 -> v1 (c947fffaa751a433c26f75ca05a03419266d098a -> c947fffaa751a433c26f75ca05a03419266d098a)
josh@brain:~/Code/numix/circle$ git status
On branch master
Your branch is up-to-date with 'origin/master'.
nothing to commit, working directory clean
josh@brain:~/Code/numix/circle$ git push --force origin master
Username for 'https://github.com': Foggalong
Password for 'https://Foggalong@github.com': 
Counting objects: 9163, done.
Delta compression using up to 4 threads.
Compressing objects: 100% (3270/3270), done.
Writing objects: 100% (9163/9163), 4.48 MiB | 1.15 MiB/s, done.
Total 9163 (delta 4708), reused 7159 (delta 4380)
To https://github.com/numixproject/numix-icon-theme-circle.git
 + f56c15d...e849d05 master -> master (forced update)
```

**Comment by wa4557** [comment]
Great that's exactly what I got :)

**Comment by Foggalong** [comment]
Excellent! We'll soon see if this fixes things then :)

**Comment by wa4557** [comment]
At least the "toxic" commits are gone:
https://github.com/numixproject/numix-icon-theme-circle/commits?page=48

**Comment by Foggalong** [comment]
\O/

**Comment by wa4557** [comment]
Hmm didn't seem to work. `git log --all -- numix-icon-theme-circle` still shows two of the bad commits, 

```
andreas@andreas-Lenovo-U530:~/Diverses/numix-icon-theme-circle$ git log --all -- numix-icon-theme-circle
commit 9c610a87e0ae850dda753c9238b85e4f6641a708
Author: Ismael González <ismgonval@gmail.com>
Date:   Sat Nov 15 16:19:41 2014 +0100

    Removed empty directory. #1288

commit c221a7007acb4142cf8946627dbd9c51d89b4dcf
Author: palob <WxWacker32@gmail.com>
Date:   Fri Nov 14 20:47:57 2014 +0100

    Knowthelist out, corrections to yate, kronometer, frtzing applied

```

which means that they are still in the history (although not in the official commits) and could be checked out. Is there a way to completele remove them?

**Comment by wa4557** [comment]
@Foggalong This is what you have to run, to completely remove every trace of the commits:

In the master branch (I forgot that the last time):

```
git for-each-ref --format='delete %(refname)' refs/original | git update-ref --stdin
git reflog expire --expire=now --all
git gc --prune=now
```
(this dereferences all the commits and garbage collects them)

And then (even more important): 

Check to the hidpi branch (because we didn't delete the bad commits there):
```
git filter-branch --force --index-filter 'git rm --cached --ignore-unmatch numix-icon-theme-circle' --prune-empty --tag-name-filter cat -- --all
```
```
 git push origin --force --tags hidpi 
```
```
git for-each-ref --format='delete %(refname)' refs/original | git update-ref --stdin
git reflog expire --expire=now --all
git gc --prune=now
```

**Comment by wa4557** [comment]
@Foggalong Can you try this (maybe before merging the last pull-request) I'm actually kind of curious if this is sufficient...

**Comment by Foggalong** [comment]
I'm actually gonna delete the HiDPi branches. I thought I'd have time to work that before university but it'll have to be put on hold for the time being. It'd also be useful to fix some of the index.theme/size stuff before we start messing with it more in HiDPi :P

**Comment by wa4557** [comment]
Cool, no more traces of the bad file, I guess the automatic updates should start working again. I'm also quite busy atm but it might get better the next few weeks :)

**Comment by palob** [comment]
Relieved to hear this is fixed finally :sweat_smile: . Thanks muchly, @wa4557.

Is the fix already applied on ``master``? The latest commits appear to be missing.

**Comment by Foggalong** [comment]
Yeah, fix applied on master :)

**Comment by wa4557** [comment]
But it doesn't seem to be enough :(

**Comment by Foggalong** [comment]
@palob is right too, we've lost a load of commits (and so icons/fixes) from the last few days too.

**Comment by Foggalong** [comment]
I just got a PPA update!

```
The following packages will be upgraded:
  numix-folders numix-icon-theme numix-icon-theme-circle
3 to upgrade, 0 to newly install, 0 to remove and 0 not to upgrade.
Need to get 9,102 kB of archives.
After this operation, 26.5 MB of additional disk space will be used.
Do you want to continue? [Y/n] Y
Get:1 http://ppa.launchpad.net/numix/ppa/ubuntu/ wily/main numix-folders all 1.1.0+201509241817~178~ubuntu15.10.1 [4,719 kB]
Get:2 http://ppa.launchpad.net/numix/ppa/ubuntu/ wily/main numix-icon-theme all 0.3+597~201509301946~ubuntu15.10.1 [3,342 kB]
Get:3 http://ppa.launchpad.net/numix/ppa/ubuntu/ wily/main numix-icon-theme-circle all 1.0.2+4~201507161646~ubuntu15.10.1 [1,041 kB]
Fetched 9,102 kB in 13s (685 kB/s)                                                          
(Reading database ... 341516 files and directories currently installed.)
Preparing to unpack .../numix-folders_1.1.0+201509241817~178~ubuntu15.10.1_all.deb ...
Unpacking numix-folders (1.1.0+201509241817~178~ubuntu15.10.1) over (1.1.0+201509241816~178~ubuntu15.04.1) ...
Preparing to unpack .../numix-icon-theme_0.3+597~201509301946~ubuntu15.10.1_all.deb ...
Unpacking numix-icon-theme (0.3+597~201509301946~ubuntu15.10.1) over (0.3+596~201509271346~ubuntu15.04.1) ...
Preparing to unpack .../numix-icon-theme-circle_1.0.2+4~201507161646~ubuntu15.10.1_all.deb ...
Unpacking numix-icon-theme-circle (1.0.2+4~201507161646~ubuntu15.10.1) over (1.0.2+4~201507161646~ubuntu15.04.1) ...
Processing triggers for hicolor-icon-theme (0.15-0ubuntu1) ...
Processing triggers for gnome-menus (3.13.3-6ubuntu1) ...
Processing triggers for desktop-file-utils (0.22-1ubuntu3) ...
Processing triggers for mime-support (3.58ubuntu1) ...
Setting up numix-folders (1.1.0+201509241817~178~ubuntu15.10.1) ...
Setting up numix-icon-theme (0.3+597~201509301946~ubuntu15.10.1) ...
Setting up numix-icon-theme-circle (1.0.2+4~201507161646~ubuntu15.10.1) ...
gtk-update-icon-cache: Cache file created successfully.
```

**Comment by ChuckDaniels87** [comment]
:tada: :tada: :tada: :tada:

**Comment by mendhak** [comment]
:trumpet: :trumpet: doot doot

**Comment by wa4557** [comment]
What? The ppa is still outdated?!

**Comment by Foggalong** [comment]
Huh, it would seem it is. How weird!

**Comment by wa4557** [comment]
@Foggalong 
Can you try this `git remote prune origin --dry-run` and post the output?

**Comment by Foggalong** [comment]
```
Pruning origin
URL: https://github.com/numixproject/numix-icon-theme-circle.git
 * [would prune] origin/hidpi
```

**Comment by wa4557** [comment]
A cool, then run it without the --dry-run :)

**Comment by Foggalong** [comment]
Done, but it's rolled back 7 days again -_-'

**Comment by wa4557** [comment]
Can you try a `git remote prune origin --dry-run` again after merging my pull-request?

**Comment by Foggalong** [comment]
Nothing, just returns :)

**Comment by wa4557** [comment]
Ok I guess this is good. However the links to the toxic commits still work, which is weird. I'm afraid this is now all I can think of, if it doesn't work now, I'm out of ideas...

**Comment by Foggalong** [comment]
no problem :)

**Comment by wa4557** [comment]
Maybe the automatic building has to be initiated again? Or is this still set up?

**Comment by Foggalong** [comment]
I think it just keeps trying automatically

**Comment by wa4557** [comment]
Ok the whole seems to work again:
https://code.launchpad.net/~paolorotolo/+junk/numix-icon-theme-circle (Look at the import details section, this is actually not the correct bzr branch btw.) so it seems that nothing is preventing the package to updated. However since the whole thing still doesn't work I guess it's something else in the packaging which I'm not experienced enough to figure out...

**Comment by Foggalong** [comment]
@PaoloRotolo ?

**Comment by bil-elmoussaoui** [comment]
I received a new update today 
`Get:36 http://ppa.launchpad.net/numix/ppa/ubuntu/ wily/main numix-icon-theme-circle all 1.0.2+4~201507161646~ubuntu15.10.1 [1 041 kB]`

**Comment by Foggalong** [comment]
@bil-elmoussaoui I think that's just the update that happens when you upgrade to 15.10 - I got the same [here](https://github.com/numixproject/numix-icon-theme-circle/issues/2430#issuecomment-144659904)

**Comment by keunes** [comment]
I am using the launchpad PPA as well, and I did receive an update (I know because it broke the Caffeine indicator #2602 :P ). Yet I have no clue if it's the relevant update: from the Ubuntu SC history I see it concerns numix-icon-theme rather than numix-icon-theme-circle, whilst I am using the latter (so I don't really understand how that would work, 'cause the Caffeine bug is only circle-related I assume)

If I could help you check if it actually is working again by providing more info, let me know how. If it wouldn't contribute anything, just ignore this comment :P

![screenshot from 2015-10-16 10 18 20](https://cloud.githubusercontent.com/assets/11229646/10537105/f0907fbe-73ef-11e5-90b6-159f6d9e05ac.png)

**Comment by dirtydancing** [comment]
Thanks for the heads-up @keunes. Numix-Circle inherits from Numix base, so you are using both :)

Unfortunately, the issue you reported at https://github.com/numixproject/numix-icon-theme-circle/issues/2609 is an issue with Numix-Circle (broken folder symlinks in Numix-Circle with target folders in Numix base that have been renamed in base due to HiDPI support).

**Comment by wa4557** [comment]
I got an update, but I guess it was manually initiated and not automatic daily build. Nevertheless issue #2608 and #2609 and #2601 are now fixed I guess

**Comment by Foggalong** [comment]
Yeah, manually initiated by Paolo to fix the issues.

**Comment by palob** [comment]
Was anyone able to figure out what's the remaining blocker here?

**Comment by wa4557** [comment]
There is actually no blocker any more. I guess the recipe for a daily build has to be applied once more...

**Comment by palob** [comment]
The PPA is 2 months behind again. If there's any issue with getting the daily build recipe up and running is there a way to have the "manual triggering" automated?

**Comment by ChuckDaniels87** [comment]
What about using a Snap package to distribute the theme in Ubuntu? I don't know if it is viable or reliable, but I think it worth trying.

**Comment by Foggalong** [comment]
The PPA isn't going to be a problem for much longer. A few minor things left to do then it'll be working again :)

**Comment by tkdberger** [comment]
>A few minor things left to do then it'll be working again :)

Aaaand still no PPA updates.

**Comment by Siilwyn** [comment]
Yes this seems to be resolved, I guess this issue can be closed right?

**Comment by palob** [comment]
Unfortunately not, it has been updated manually a few weeks ago.

**Comment by Foggalong** [comment]
Updated - despite the repo switched and new build system this still doesn't appear to be fixed.

**Comment by wa4557** [comment]
I guess this is fixed now :)

**Comment by tkdberger** [comment]
I'm cautiously optimistic.

**Comment by mrmeszaros** [comment]
Hi, I think the issue is still open, since:

1. My PR for Terraria icon (#3635) has been merged on Apr 12.

2. I have received updates to the `numix-icon-theme-circle` package

3. but I do **NOT** see the icon in the `numix-icon-theme-circle` package updates

4. So I downloaded the sources from the PPA: first enabling sources, then running `apt-get source numix-icon-theme-circle`, and got `numix-icon-theme-circle_2.0.3+14~201704051631~ubuntu16.04.1.tar.xz`

5. The name of the package suggests that it was created on April 5th 2017.

6. Checked at [Launchpad](https://launchpad.net/~numix/+archive/ubuntu/ppa) - it also shows _Numix Maintainers (2017-04-05)_

Why are the circle icon updates released so rarely?
